apt_key supporting proxy(#31691)

hangsu-ma opened the pull request:
Fixes https://github.com/ansible/ansible/issues/31691

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->
apt-key does not respect Acquire::http::Proxy specified in apt conf files, nor respect http_proxy environment variable.
More discussion about these behaviours can be found here: https://bugs.launchpad.net/ubuntu/+source/software-properties/+bug/1433761

keyserver-options are used to pass in proxy settings for apt-key, example:
sudo apt-key adv --keyserver-options http-proxy=http://username:password@proxy.example.com:8080 --keyserver keyserver.ubuntu.com --recv-keys GPG_KEY

This fix adds a new option for apt_key module to support keyserver-options.

It also enable apt-key to read proxy settings set in environments. This is not consistent with apt-key command line behaviour, but hey, ansible is suppose to make things easier. And it does make this module in-line with other modules in terms of proxy settings, and made it possible to put this module in same block with other modules and have the proxy set on block level.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
apt_key

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
all
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->
<!--- Paste verbatim command output below, e.g. before and after your change -->
```

WARNING: apt-key is deprecated according to this: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=851774
This fix is made for backward compatibility only, so existing playbook can work behind proxy.
```

jamescassell commented on lib/ansible/modules/packaging/os/apt_key.py:
```diff
@@ -211,11 +229,19 @@ def download_key(module, url):
 
-def import_key(module, keyring, keyserver, key_id):
+def import_key(module, keyring, keyserver, keyserver_options, key_id):
+    if (environ.get('http_proxy')):
```
If you're supporting http_proxy, you should also support no_proxy. Is https_proxy also relevant here? Seems like it would be if a key server could ever be https.

hangsu-ma replied:
I actually have both https_proxy and no_proxy in the code, had to took them out pretty quickly, because they are not supported options, see error message below if you pass them in from --keyserver-options:
```
sudo apt-key adv --keyserver-options "http-proxy=http://username:password@proxy.example.com:8080 https-proxy=http://username:password@proxy.example.com:8080 https_proxy=http://username:password@proxy.example.com:8080 no-proxy=www.google.com no_proxy=www.google.com" --keyserver keyserver.ubuntu.com --recv-keys GPG_KEY
Executing: /tmp/apt-key-gpghome.zJ3uDH1oNy/gpg.1.sh --keyserver-options http-proxy=http://username:password@proxy.example.com:8080 https-proxy=http://username:password@proxy.example.com:8080 https_proxy=http://username:password@proxy.example.com:8080 no-proxy=www.google.com no_proxy=www.google.com --keyserver keyserver.ubuntu.com --recv-keys GPG_KEY
gpg: keyserver option 'https-proxy' is unknown
gpg: keyserver option 'https_proxy' is unknown
gpg: keyserver option 'no-proxy' is unknown
gpg: keyserver option 'no_proxy' is unknown
gpg: "GPG_KEY" not a key ID: skipping

```
below are my speculation due to the lack of documentation on keyserver-options:
* no proxy, this apt-key is hitting single URL(the key server) which is either in the proxy or outside the proxy, so no_proxy or such is needed. I guess apt-key will fail when there is a URL forward or redirect on key server that cross proxy (not sure if it's allowed by OpenPGP)
* https proxy, according to the answers on this: https://security.stackexchange.com/questions/4161/shouldnt-gpg-key-fetching-use-a-secure-connection
it's always on http protocal so https_proxy is not releveant and not supported.

jamescassell replied:
In that case, properly supporting no_proxy would require parsing whether any domain suffix in no_proxy is a suffix of the keyserver requested.  If it is, the http-proxy option should not be passed.

hangsu-ma replied:
Hi @jamescassell 
I have implemented the suggested logic

jamescassell reviewed (changes requested):
Looks like a good change overall. I would like to also see support for no_proxy, otherwise risk breaking current usage....

jamescassell reviewed (changes requested):
Adding support for http_proxy without no_proxy support risks breaking existing setups where the keyserver is in the no_proxy list.

weber-d commented:
I have the same problem: Want to add the docker registry to my play using `apt_key`
```
- name: Add docker GPG key
    apt_key:
      url: https://download.docker.com/linux/ubuntu/gpg
      state: present
```
Sadly, it seems that no solution exists. So I'm fored to use plain commands, altough they're considered as bad practice in Ansible when dedicated commands are avaliable.

jamescassell commented on lib/ansible/modules/packaging/os/apt_key.py:
```diff
@@ -128,6 +146,51 @@ def find_needed_binaries(module):
+            return True
+        elif u.startswith('.') and k.endswith(u):
+            return True
+        elif u.endswith('.') and k.startswith(u):
```
I'd drop this feature, as I'm not aware of any other no_proxy processing that does prefix matching.

jamescassell commented on lib/ansible/modules/packaging/os/apt_key.py:
```diff
@@ -128,6 +146,51 @@ def find_needed_binaries(module):
+                from ansible.module_utils._text import to_text
+                if ipaddress.ip_address(to_text(k)) in ipaddress.ip_network(to_text(u)):
+                    return True
+            except ImportError as err:
```
For this case and the one below, I would just `pass` and assume it's not a match.  Most implementations of `no_proxy` don't support CIDR matching.

jamescassell reviewed (changes requested):
shipit

Couple of tweaks requested.

gopinatht commented:
+1 for this PR. Is there a timeline for merging it?

egroeper commented:
@hangsu-ma Still interested in merging this? Or should / may somebody else step in?

Would really like to see this feature.

SarahSamji commented:
When is this feature going to be made available? Or is there a workaround to adding http_proxy and no_proxy?

hangsu-ma commented:
> @hangsu-ma Still interested in merging this? Or should / may somebody else step in?
> 
> Would really like to see this feature.
@egroeper i am back on this now, will need to pick it up again and go through latest comments then test with new Ubuntu releases, maybe Ubuntu/Debien fixed it already. Will update this PR.

mohag commented:
GPG, used by apt-key, has an option [`--honor-http-proxy` since at least 2004](https://github.com/gpg/gnupg/blob/592deeddb9bf4ae9b3e236b439e2f39644eb6d46/dirmngr/ONEWS#L141). (Not documented in the man page...)

[http-proxy without arguments also results in the environment variable being used](https://github.com/gpg/gnupg/blob/592deeddb9bf4ae9b3e236b439e2f39644eb6d46/dirmngr/ChangeLog-2011-ks#L532). (At least at some point)

[This](https://www.gnupg.org/documentation/manuals/gnupg/GPG-Configuration-Options.html) might also be interesting. (Search for http-proxy on the page) (Ubuntu 18.04 has GnuPG 2.2.4) (I'm not sure if the options gets passed to right places...)

If one of the options to make GnuPG look at the environment variables are used, the lack of `no_proxy` support is then GPG's bug, but a workaround in the module might still be useful...

mohag commented:
The other approach might be to allow the user to pass extra options from the module...

hangsu-ma committed:
check if a given host is included in the given no_proxy list

So that the caller can determine if proxy settings should be applied.
Supports most common form of no_proxy listed below:
1. no_proxy is a comma separated list of one or more items below
2. full domain name or hostname
3. containing * wildcard. example: *.example.com
4. leading or trailing dot (.). example: 192.168.56.
5. subnet range. example: 192.168.23.0/24

NOTE: no DNS look up is carried out

hangsu-ma commented:
@jamescassell @bcoca 

The problem still persists on ubuntu 18.04, 19.04 and 19.10. I have tested the fixes on them as well using this [project](https://github.com/hangsu-ma/ansible-proxy-sandbox).

I have "rebased" the change to latest code base, resulted in a force push sadly.
* implemented changes @jamescassell recommended
* fixed a NPL when no_proxy is not specified. 
* shared code between this one and https://github.com/ansible/ansible/pull/42536 are also moved into module_utils. 

The [other one](https://github.com/ansible/ansible/pull/42536) has less love it seems. Can you have a look at both PRs and let me know if anything else is needed?

most of my time on this has been spent on setting up sandbox with no internet access but only a proxy in vagrant, so i can work on it on my own PC. Which results in [this](https://github.com/hangsu-ma/ansible-proxy-sandbox), for anyone want to work on these without access to cooperate proxy in office, you can have a look.

hangsu-ma committed:
adding proxy support for apt_key (#31691)

WARNING: apt-key is deprecated according to this: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=851774
This fix is made for backward compatibility only, so existing playbook can work behind proxy.

apt-key does not respect Acquire::http::Proxy specified in apt conf files, nor http_proxy environment variable.
More discussion about these behaviours can be found here: https://bugs.launchpad.net/ubuntu/+source/software-properties/+bug/1433761

keyserver-options are used to pass in proxy settings for apt-key, example:
sudo apt-key adv --keyserver-options http-proxy=http://username:password@proxy.example.com:8080 --keyserver keyserver.ubuntu.com --recv-keys GPG_KEY

The change enables apt-key to read proxy settings set in environments variable http_proxy and no_proxy. This is not consistent with apt-key command line behaviour (as in proxy needs to be passed in using keyserver-options for apt-key), but does make this module in-line with other modules in terms of proxy settings, and made it possible to put this module in same block with other modules and have the proxy set on block level. Ansible is suppose to make things easier after all.

hangsu-ma commented:
shipit

hangsu-ma commented:
build failing on create VM, anyway to restart it?

mkrizek reviewed (changes requested):
To prevent accidental merging of this I am changing the review to `Request changes` per the last comment from ansibot.

mkrizek commented:
cc @mattclay

hangsu-ma commented:
@mkrizek 
Hi Martin, i didn't find a collection(os_migrate.os_migrate in the bots comment) contains apt_key.py. It's still on devel branch in main ansible repo. Can you point me to the correct place to create a new PR?
For network.py, there is a new method added in, tests seems only using the version in test folder and will be failing if i dont add the same method in that folder. What is the correct approach for these cases?

mkrizek commented:
> @mkrizek
> Hi Martin, i didn't find a collection(os_migrate.os_migrate in the bots comment) contains apt_key.py. It's still on devel branch in main ansible repo. Can you point me to the correct place to create a new PR?

The message mentions `test/support/integration/plugins/module_utils/common/network.py`, not `apt_key.py`.

> For network.py, there is a new method added in, tests seems only using the version in test folder and will be failing if i dont add the same method in that folder. What is the correct approach for these cases?

Generally files in test/support should not be modified. `lib/ansible/module_utils/common/network.py` and `test/support/integration/plugins/module_utils/common/network.py` appear to be of same content. I pinged @mattclay to guide you through this specific case.

hangsu-ma commented:
Hi @mkrizek
thanks for the swift reply, the only network.py i can find is 
https://github.com/ansible-collections/ansible.netcommon/blob/master/plugins/module_utils/network/common/network.py
But it seems to be a different file with same name, maybe the bot is recognizing the file only by name, and getting a false positive?

mkrizek commented:
> Hi @mkrizek
> thanks for the swift reply, the only network.py i can find is
> https://github.com/ansible-collections/ansible.netcommon/blob/master/plugins/module_utils/network/common/network.py
> But it seems to be a different file with same name, maybe the bot is recognizing the file only by name, and getting a false positive?

No, the bot looks for files in `test/support` and warns if someone attempts to modify them in a pull request which is correct in this case. Then it is on case by case basis for a human to determine if the change in `test/support` is desired or not.

hangsu-ma commented:
> 
> No, the bot looks for files in `test/support` and warns if someone attempts to modify them in a pull request which is correct in this case. Then it is on case by case basis for a human to determine if the change in `test/support` is desired or not.

Cool, i will be waiting for further instructions on this case then @mkrizek @mattclay.
There was a commit without changes in test/support/integration/plugins/module_utils/common/network.py, but test cases are failing to find the new method added in network.py, it seems to loadi test/support/integration/plugins/module_utils/common/network.py in preference to lib/ansible/module_utils/common/network.py. Took me a while to figure it out and had to do the same change in test/support/integration/plugins/module_utils/common/network.py to make the test pass. That particular commit is long gone now after few rebase, but should be easy to reproduce.

mattclay commented:
@hangsu-ma The file `test/support/integration/plugins/module_utils/common/network.py` has been removed from the repo, so this PR needs to be rebased.

hangsu-ma commented:
> @hangsu-ma The file `test/support/integration/plugins/module_utils/common/network.py` has been removed from the repo, so this PR needs to be rebased.

@mattclay thanks for the fix, i have now rebased again.

chrisz100 commented:
Any update on this?

bcoca commented:
proxy suport was added to the internal `fetch_file` function that apt_key uses so this is fixed in newer versions.

KawaiDesu commented:
> this is fixed in newer versions.

Which is "newer"? Newer than 2.7?

mohag commented:
> > this is fixed in newer versions.
> 
> Which is "newer"? Newer than 2.7?

`fetch_file` is only used if a URL is used, if a key ID is give, it runs apt-key, which does not use the proxy settings...

[An easy workaround](https://github.com/ansible/ansible/pull/42443#issuecomment-574552647) is to just build the URL directly.
